src/Init/Lean/Elab/Quotation.lean

```diff
@@ -101,10 +107,10 @@ private partial def quoteSyntax {m : Type → Type} [Monad m] [MonadQuotation m]
 | _, Syntax.missing => unreachable!
 
 def stxQuot.expand {m : Type → Type} [Monad m] [MonadQuotation m] (env : Environment) (stx : Syntax) : m Syntax := do
-  -- TODO: hygiene
-  -- `(do msc ← getCurMacroScope; pure %(quote `(msc) quoted))
-  stx ← quoteSyntax env Syntax.missing 0 $ stx.getArg 1;
-  `(HasPure.pure %%stx)
+let quoted := stx.getArg 1;
+-- `(do msc ← getCurrMacroScope; pure %(quoteSyntax env 0 quoted))
+stx ← quoteSyntax env 0 quoted;
+`(HasBind.bind Lean.MonadQuotation.getCurrMacroScope (fun scp => HasPure.pure %%stx))
```

Comment:
@leodemoura `getCurrMacroScope` is now being used here; see also the test output below for its effects.

Reply:
Comments, please! This is clever, but it is not obvious at all. I will forget how this works in a few months.
Suggestions: refresh the reader's memory that 
- our quotations are effectful, and rely on `getCurrMacroScope`. 
- `quoteSyntax` generates code that uses `scp`.

Reply:
BTW, `scp` can be accidentally captured. Example:
```lean
#eval run $ do scp ← `(Nat.one); `(%%scp)
```
fails for me.

Reply:
Here is the error message I get:
```lean
/Users/leonardodemoura/projects/lean4/tests/lean/StxQuot.lean:15:0: error: type mismatch at application
  pure scp
term
  scp
has type
  MacroScope
but is expected to have type
  Syntax
```

Reply:
Now with comments. I've tried to mostly comment the implementation and keep the general hygiene approach to the paper, but I can also copy a bit more paper content into comments if you want.

> BTW, scp can be accidentally captured.

Yes, but the next update-stage0 will fix that: the code will change *its own behavior* so that the identifier `scp` is now introduced with a macro scope :) .

Reply:
We need more comments. Using quotations to implement quotations here is making the code very hard to understand. We need at least a comment explaining how using `scp` in different quotations do not break hygiene.
To be honest, I think the code would be much easier to understand if it was implemented without quotations. Then, all the magic would be clear to the code reviewer.